DM-41365: Add findDataset/getDataset/getDatasetType to Butler API

timj opened the pull request:
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

timj committed:
Add DirectButler.get_dataset_type API

This replaces butler.registry.getDatasetType

timj committed:
Add get_dataset_type to RemoteButler

timj committed:
Add DirectButler.find_dataset

timj committed:
Add Butler.get_dataset

timj committed:
Add news fragment

timj committed:
Check whether remote butler thinks it is writeable

timj commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -772,6 +773,126 @@ def getURI(
+    @abstractmethod
+    def find_dataset(
+        self,
+        datasetType: DatasetType | str,
```
@TallJimbo I assume we aren't going to start saying `dataset_type` and `data_id` in the public APIs?

TallJimbo replied:
I figured we _would_ start using snake_case for those, too, when they're new arguments or new methods.  I've already started doing that in various places.

dhirving commented on tests/test_server.py:
```diff
@@ -77,6 +82,10 @@ def create_factory_dependency():
         cls.client = TestClient(app)
         cls.butler = _make_remote_butler(cls.client)
 
+        # Populate the test server.
+        server_butler.import_(filename=os.path.join(TESTDIR, "data", "registry", "base.yaml"))
+        server_butler.import_(filename=os.path.join(TESTDIR, "data", "registry", "datasets-uuid.yaml"))
+
```
Doesn't matter yet but you may want to move this above the call to _make_remote_butler, in case we decide to read some stuff from the server during RemoteButler initialization.

timj replied:
I'm hoping that it doesn't make a difference here either way because the clients won't be downloading caches of dataset type and collection before this runs will they?

TallJimbo commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -798,6 +799,83 @@ def get_dataset_type(self, name: str) -> DatasetType:
+        datastore_records: bool = False,
+        **kwargs: Any,
+    ) -> DatasetRef | None:
+        """Find a dataset given its `DatasetType` and data ID.
```
With this method moving to `Butler`, I think it'd make sense for it to become an entry point for the logic that handles `day_obs` and `seq_num` (etc) as data ID keys.  I think it has everything it needs.

timj replied:
Okay. This can in theory call `Butler._rewrite_data_id` now that it's moved up to Butler.

timj replied:
I've done this and it works fine. It doesn't save us anything in `butler.get()` though because that does the day_obs/seq_num look up early so it can construct a proper DataId, and potentially expand it, before calling find_dataset.

TallJimbo commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -798,6 +799,83 @@ def get_dataset_type(self, name: str) -> DatasetType:
+        *,
+        collections: CollectionArgType | None = None,
+        timespan: Timespan | None = None,
+        datastore_records: bool = False,
```
If we provide this as an option, we should also provide a way to expand the data ID (i.e. add dimension records) at the same time, and we should think about whether we want just one option to do both or allow the two kinds of expansion to happen separately.

timj replied:
I had completely failed to notice that that flag was in there now. One option is for `find_dataset` to always expand and add everything. People shouldn't be calling it in bulk.

andy-slac replied:
Always expanding could be inefficient, I think many clients do not care about datastore records at all, we should not be pessimizing things for them.

timj replied:
Presumably get_dataset should have the datastore_records and dimension_records booleans as well for consistency?

TallJimbo replied:
:+1:

timj replied:
I also need to add datastore records to the SerializedDatasetRef because that was never updated and so it's impossible to return datastore records from the server at the moment...

andy-slac replied:
Jim asked me to not add them to `SerializedDatasetRef`, as it may increase graph size on disk (but I don't think we give records to refs that go into the graph, at least not yet).

timj replied:
We are sending SerializedDatasetRef over the wire so we can only reconstruct things in the far end that were put in at the other end. We need to add them somehow (but obviously they should be optional). For now I've turned the feature off in client/server so you can't get the datastore records.

TallJimbo commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -798,6 +799,83 @@ def get_dataset_type(self, name: str) -> DatasetType:
+        datasetType: DatasetType | str,
+        dataId: DataId | None = None,
+        *,
+        collections: CollectionArgType | None = None,
```
This needs to be an ordered search path, so some of the types in this alias like `re.Pattern` and `EllipsisType` aren't actually valid here.

timj replied:
So really it's `str | Iterable[str] | CollectionSearch`  and the original annotation isn't quite right.

TallJimbo replied:
Yes, and I think `CollectionSearch` is deprecated (and ready to be removed).

TallJimbo commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -798,6 +799,83 @@ def get_dataset_type(self, name: str) -> DatasetType:
+        collections: CollectionArgType | None = None,
+        timespan: Timespan | None = None,
+        datastore_records: bool = False,
+        **kwargs: Any,
```
Should we consider making this accept `storage_class`, too, so storage-class-overrides in `get` (etc) can be delegated here and then passed with the `ref` instead of separately?

timj replied:
Can give that a try.

TallJimbo commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -798,6 +799,83 @@ def get_dataset_type(self, name: str) -> DatasetType:
+            ``self.collections`` is `None`.
+        LookupError
+            Raised if one or more data ID keys are missing.
+        lsst.daf.butler.registry.MissingDatasetTypeError
```
We need to move this exception out of `registry` now.

timj replied:
I had in fact done this (for now I'm lifting it out) but I forgot to update the docs.

TallJimbo commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -799,6 +799,23 @@ def get_dataset_type(self, name: str) -> DatasetType:
         raise NotImplementedError()
 
+    @abstractmethod
+    def get_dataset(self, id: DatasetId) -> DatasetRef | None:
```
I think we'll want a vectorized version of this in the future, but it's a pain to implement now.  I don't think this interface needs to change in order to live alongside a vectorized one, but I wanted to raise it in case you think otherwise.

timj replied:
There are two open questions:

* If we have vectorized version, `get_datasets` do we keep `get_dataset` around?
* What should `get_datasets` return? `list[DatasetRef]` or `DatasetRefContainer`?

I think we likely want a single request to return a single ref so it seems like we always want `Butler.get_dataset` even if we later add `Butler.get_datasets`.

andy-slac replied:
Hmm, do we ever expose bare `DatasetId` to butler clients (do we need this method at butler level)?

timj replied:
It certainly makes client/server testing a lot easier to begin with since it's the simplest possible API to implement. We do have VO services that are given a UUID (via datalink from, eg, ObsTAP) and need to convert that to a dataset ref so it's not impossible that the server will need to do that (depending on how services connect to butlers).

TallJimbo replied:
I think we'll want to start exposing dataset IDs in more places, too; I've been thinking about how we might to get them into the `--where` expression language, for instance. They appear pretty prominently in QGs already.

TallJimbo reviewed (approved):
Just took a look at the APIs, and while I left a lot of comments they're all minor.  Did not look at the implementations (I gather that's not needed from me).

andy-slac commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -772,6 +773,126 @@ def getURI(
+        Notes
+        -----
+        This method handles component dataset types automatically, though most
+        other registry operations do not.
```
Drop `registry` here or replace with `Butler`?

andy-slac commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -772,6 +773,126 @@ def getURI(
+        **kwargs
+            Additional keyword arguments passed to
+            `DataCoordinate.standardize` to convert ``dataId`` to a true
+            `DataCoordinate` or augment an existing one.
```
`datastore_records` needs to be added to docstring, somehow I forgot to add it to the registry method when introduced it first.

andy-slac commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -772,6 +773,126 @@ def getURI(
+        never changes the behavior.
+
+        This method handles component dataset types automatically, though most
+        other registry operations do not.
```
"registry" here too.

andy-slac commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -42,15 +42,16 @@
 from .datastore import DatasetRefURIs, Datastore
 from .dimensions import DataId, DimensionConfig
-from .registry import Registry, RegistryConfig, _RegistryFactory
+from .registry import CollectionArgType, Registry, RegistryConfig, _RegistryFactory
```
We should probably move `CollectionArgType` out of `registry`. I think ideally `Butler` should not depend on anything from `registry` (we keep `Butler.makeRepo()` so we need three other classes here.)

timj replied:
I replaced it with an explicit `str | Sequence[str]` definition.

andy-slac commented on python/lsst/daf/butler/remote_butler/_remote_butler.py:
```diff
@@ -179,6 +211,57 @@ def getURI(
+        # and only go to the server if the dataset type is not known.
+        path = f"dataset_type/{name}"
+        response = self._client.get(self._get_url(path))
+        response.raise_for_status()
```
Will this raise some HttpError instead of `MissingDatasetTypeError` when dataset type is not know?

andy-slac commented on python/lsst/daf/butler/remote_butler/_remote_butler.py:
```diff
@@ -179,6 +211,57 @@ def getURI(
+    def get_dataset(self, id: DatasetId) -> DatasetRef | None:
+        path = f"dataset/{id}"
+        response = self._client.get(self._get_url(path))
+        if response.json() is None:
```
Does this need some error handling before you try to covert response to JSON (`raise_for_status()`?)

timj replied:
Yes. Oops.

timj replied:
Turns out FastAPI exception handling is an adventure (you have to raise an HttpException and fill it with all the real exception information and then re-raise in the client). I think I'm going to punt on this for now.

andy-slac commented on python/lsst/daf/butler/remote_butler/_remote_butler.py:
```diff
@@ -179,6 +211,57 @@ def getURI(
+                    "No collections provided to find_dataset, and no defaults from butler construction."
+                )
+            collections = self.collections
+        # Temporary hack. Assume strings for collections. In future
+        # want to construct CollectionWildcard and filter it through collection
+        # cache to generate list of collection names.
```
Working on caching ticket now, I would probably prefer to not have a complete list of collections on every client in the long term.

timj replied:
Interesting. I thought we said we would last week. I thought @TallJimbo wanted the client to know all possible collections when it was building the query object before sending it to the server. The caching story between client and server is a bit tricky.

andy-slac replied:
Yes, I know, but we may need to rethink that part, otherwise we may have scaling issue. I looked at Butler initialization time today, the simplest `butler query-data-ids --collections=HSC/raw/all /repo/main exposure` takes 10 seconds, large fraction of that comes from pre-loading of collections cache. We have 300k collections today, and it will only grow with time. There are ways to reduce that probably, but in my ideal world it would be better if we cache almost nothing on client side. 🙂

TallJimbo replied:
I think having the caches in the client has worked well for us so far (i.e. with the server just being the DB), and I like the idea of being able to optimize server calls away entirely in the client sometimes.  And I do think caching on the server is a lot harder, since we can't just punt cache misses to the user there.

But I do hesitate a bit because I'm not sure if "cache all the collections the user can see" continues to be in the "small enough to be irrelevant" category after DR1, both in terms of downloading the cache and holding it in memory.  And I have no idea if my "do more on the client to keep the server load light" intuition is better than @dhirving 's "do nothing on the client to keep version management easier" intuition.

andy-slac replied:
Downloading a million collection records just to figure out that pattern does not match any of them looks like a lot of overhead. It is of course about the right balance between what we do on server vs client, we need to find something that works and scales.

TallJimbo replied:
Ok, I'm sold on not trying to aggressively fetch all collections the user can access.  It's the dataset type definitions, chained collection members, and collection summaries that are needed for query contruction, and we don't need all of any of those.

I think what may break if we drop the complete collection-name cache is our regex/glob collection query implementation - we'll need to start translating/limiting those patterns to something that can be run inside the DB.  Yes, we could do a full cache in the `RemoteButler` server, but this scaling problem will hit `DirectButler`, too.

I'll do a brain dump on Slack later today about the places we use the caches in the query system so somebody other than me can do some conceptual design work on what it should look like.  I think we need to have a better sense of where this is all going, even if we don't decide to switch fully from the current approach yet, but I want to stay off of at least this part of the critical path.

andy-slac commented on python/lsst/daf/butler/remote_butler/_remote_butler.py:
```diff
@@ -179,6 +211,57 @@ def getURI(
+        # Temporary hack. Assume strings for collections. In future
+        # want to construct CollectionWildcard and filter it through collection
+        # cache to generate list of collection names.
+        wildcards = CollectionWildcard.from_expression(collections)
```
Maybe do a quick check here and raise if `patterns` specifies something?

timj replied:
Seems like I have to change the API to something a bit simpler for collection list anyhow.

andy-slac commented on python/lsst/daf/butler/remote_butler/server/_server.py:
```diff
@@ -56,8 +64,88 @@ def factory_dependency() -> Factory:
+) -> SerializedDatasetType:
+    """Return the dataset type."""
+    butler = factory.create_butler()
+    datasetType = butler.get_dataset_type(dataset_type_name)
```
Is exception transported transparently back to client?

andy-slac commented on tests/test_server.py:
```diff
@@ -90,6 +99,38 @@ def test_simple(self):
+
+    def test_get_dataset_type(self):
+        bias_type = self.butler.get_dataset_type("bias")
+        self.assertEqual(bias_type.name, "bias")
```
Add a check for non-existent one?

andy-slac reviewed (approved):
Looks good, few minor questions.

timj committed:
Use data_id/dataset_type in find_dataset and simplify collections

The underscores are the preferred way to add new APIs now.

The collections parameter should not support wildcards so explicitly
declare it should be a sequence of str.

timj committed:
Add support for day_obs/seq_num to Butler.find_dataset

timj committed:
Add storage class conversion to get_dataset and find_dataset

timj committed:
Remove lifting of CollectionArgType

timj committed:
Add missing instrument metadata to test yaml

visit_system is now required but no test was triggering
validation until recently.

timj committed:
Add dimension and datastore record retrieval for get_dataset/find_dataset

Datastore records do not work in remote butler because they
were never added to SerializedDatasetRef.

timj committed:
Add experimental exception handling in client/server

Demonstrate that MissingDatasetTypeError can be caught in
the server and raised again in the client.
